exchange/exchange.go

```diff
@@ -267,7 +269,7 @@ func (e *exchange) HoldAuction(ctx context.Context, r *AuctionRequest, debugLog
 	conversions := e.getAuctionCurrencyRates(requestExtPrebid.CurrencyConversions)
 
 	if e.floor.Enabled {
-		floorErrs = floors.EnrichWithPriceFloors(r.BidRequestWrapper, r.Account, conversions)
+		floorErrs = floors.EnrichWithPriceFloors(r.BidRequestWrapper, r.Account, conversions, e.priceFloorFetcher)
```

Comment:
If `e.floor.Enabled` evaluates to `false`, we don't do price floors anywhere in the code at all, right? It seems that we don't fetch for price floors nor we check for price floors in the bid request itself neither here nor in line 389 below. Therefore, I don't think we need to initialize the fetcher at all if `e.floor.Enabled` is `false`. `e.floor.Enabled` is a host config value that doesn't change from request to request. If `e.floor.Enabled` is false, we could optimize a bit by not initializing the `PriceFloorFetcher` that comes with `Ticker` and channels overhead. Maybe we could check for this value back in `router.go` and pass a `nil` `PriceFloorFetcher` to the `Exchange` if `e.floor.Enabled` is `false`:
```
227
    +   var priceFloorFetcher *PriceFloorFetcher = nil
    +   if cfg.PriceFloors.Enabled {
228         priceFloorFetcher := floors.NewPriceFloorFetcher(cfg.PriceFloors.Fetcher.Worker, cfg.PriceFloors.Fetcher.Capacity,
229             cfg.PriceFloors.Fetcher.CacheSize, floorFechterHttpClient, r.MetricsEngine)
    +   }
230
231     tmaxAdjustments := exchange.ProcessTMaxAdjustments(cfg.TmaxAdjustments)
232     planBuilder := hooks.NewExecutionPlanBuilder(cfg.Hooks, repo)
233     macroReplacer := macros.NewStringIndexBasedReplacer()
234     theExchange := exchange.NewExchange(adapters, cacheClient, cfg, syncersByBidder, r.MetricsEngine, cfg.BidderInfos, gdprPermsBuilder, rateConvertor, categoriesFetcher, adsCertSigner
235     var uuidGenerator uuidutil.UUIDRandomGenerator
router/router.go
```
We could do a `nil` check here instead: 
```
231   func (e *exchange) HoldAuction(ctx context.Context, r *AuctionRequest, debugLog *DebugLog) (*AuctionResponse, error) {
232       if r == nil {
233           return nil, nil
234       }
235 *-- 35 lines: var floorErrs []error------------------------------------------------------------------------------------------------------------------------------------------------------
270  
271 -     if e.floor.Enabled {
    +     if priceFloorFetcher != nil {
272           floorErrs = floors.EnrichWithPriceFloors(r.BidRequestWrapper, r.Account, conversions, e.priceFloorFetcher)
273       }
exchange/exchange.go
```
Thoughts?

Reply:
Sure, will modify the code
Also i think we don't need this check again
```go
if priceFloorFetcher != nil {
       floorErrs = floors.EnrichWithPriceFloors(r.BidRequestWrapper, r.Account, conversions, e.priceFloorFetcher)
}
```
if floors enabled is `true`, then we are creating the floor fetcher instance, so we will not get condition where floors enabled is `true` and fetcher instance is `nil`

Reply:
Instead for SafeGuard i will add check when we actually use fetcher instance
```
 func (f *PriceFloorFetcher) Fetch(config config.AccountPriceFloors) (*openrtb_ext.PriceFloorRules, string) {
-       if !config.UseDynamicData || len(config.Fetcher.URL) == 0 || !validator.IsURL(config.Fetcher.URL) {
+       if f == nil || !config.UseDynamicData || len(config.Fetcher.URL) == 0 || !validator.IsURL(config.Fetcher.URL) {
                return nil, openrtb_ext.FetchNone
        }
```

Reply:
Also
Instead of writing the below logic in `router.go`
```
 var priceFloorFetcher *PriceFloorFetcher = nil
    +   if cfg.PriceFloors.Enabled {
228         priceFloorFetcher := floors.NewPriceFloorFetcher(cfg.PriceFloors.Fetcher.Worker, cfg.PriceFloors.Fetcher.Capacity,
229             cfg.PriceFloors.Fetcher.CacheSize, floorFechterHttpClient, r.MetricsEngine)
    +   }
```
The internal implementation of creating fetcher instance will take care of it
```
-func NewPriceFloorFetcher(maxWorkers, maxCapacity, cacheSize int, httpClient *http.Client, metricEngine metrics.MetricsEngine) *PriceFloorFetcher {
+func NewPriceFloorFetcher(config config.PriceFloors, httpClient *http.Client, metricEngine metrics.MetricsEngine) *PriceFloorFetcher {
+       if !config.Enabled {
+               return nil
+       }
+
        floorFetcher := PriceFloorFetcher{
-               pool:            pond.New(maxWorkers, maxCapacity, pond.PanicHandler(workerPanicHandler)),
+               pool:            pond.New(config.Fetcher.Worker, config.Fetcher.Capacity, pond.PanicHandler(workerPanicHandler)),
                fetchQueue:      make(FetchQueue, 0, 100),
                fetchInProgress: make(map[string]bool),
-               configReceiver:  make(chan FetchInfo, maxCapacity),
+               configReceiver:  make(chan FetchInfo, config.Fetcher.Capacity),
                done:            make(chan struct{}),
-               cache:           freecache.NewCache(cacheSize * 1024 * 1024),
+               cache:           freecache.NewCache(config.Fetcher.CacheSize * 1024 * 1024),
                httpClient:      httpClient,
                time:            &timeutil.RealTime{},
                metricEngine:    metricEngine,
```